op-supervisor/supervisor/backend/reads/full.go

```diff
@@ -0,0 +1,191 @@
+package reads
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+// readHandle maintains a lower and upper range view over a chain.
+// This is invalidated through the Registry whenever the chain changes.
+type readHandle struct {
+
+	// highestDerivedTimestamp: highest timestamp that this read depends on.
+	// Set to block number that the read handle depends on.
+	// This may be changed at any time, to adjust the read-handle.
+	// If nil, then there is no dependency on derived time.
+	highestDerivedTimestamp *uint64
+
+	// lowestInvalidatedDerivedTimestamp: lowest timestamp that was invalidated while this read handle was active.
+	// If nil, then nothing was invalidated.
+	lowestInvalidatedDerivedTimestamp *uint64
+
+	highestSourceNum           *uint64
+	lowestInvalidatedSourceNum *uint64
+
+	// Note: we may add a lowestDependencyNum and a highestPrunedNum in the future, to handle DB pruning.
+
+	// the read handle is locked when inspecting it and making changes to it
+	mu sync.Mutex
+
+	registry *Registry
+}
+
+func (h *readHandle) DependOnDerivedTime(timestamp uint64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.registry == nil {
+		panic(fmt.Errorf("cannot depend on derived timestamp %d, read-handle has already been released", timestamp))
+	}
+	prev := h.highestDerivedTimestamp
+	if prev == nil || timestamp > *prev {
+		h.highestDerivedTimestamp = &timestamp
+	}
+}
+
+func (h *readHandle) DependOnSourceBlock(blockNum uint64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.registry == nil {
+		panic(fmt.Errorf("cannot depend on source block %d, read-handle has already been released", blockNum))
+	}
+	prev := h.highestSourceNum
+	if prev == nil || blockNum > *prev {
+		h.highestSourceNum = &blockNum
+	}
+}
+
+func (h *readHandle) invalidateDerived(timestamp uint64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.registry == nil { // if already released, then we can ignore the invalidation signal
```

Comment:
Rather than checking if the handle if it is closed all the time, could the handle just replace itself when closed with a no-op handle? Not sure this is a great idea, just thinking out loud.

Reply:
Changing the type of the method receiver with some indirection creates more edge cases than it's worth I think